File: pathplannerlib/src/main/java/com/pathplanner/lib/auto/CommandUtil.java
package com.pathplanner.lib.auto;

import com.pathplanner.lib.path.PathPlannerPath;
import edu.wpi.first.wpilibj2.command.*;
import java.io.IOException;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.ParseException;

/** Utility class for building commands used in autos */
public class CommandUtil {
  /**
   * Wraps a command with a functional command that calls the command's initialize, execute, end,
   * and isFinished methods. This allows a command in the event map to be reused multiple times in
   * different command groups
   *
   * @param eventCommand the command to wrap
   * @return a functional command that wraps the given command
   */
  public static Command wrappedEventCommand(Command eventCommand) {
    return new FunctionalCommand(
        eventCommand::initialize,
        eventCommand::execute,
        eventCommand::end,
        eventCommand::isFinished,
        eventCommand.getRequirements().toArray(Subsystem[]::new));
  }

  /**
   * Builds a command from the given JSON object.
   *
   * @param commandJson the JSON object to build the command from
   * @param loadChoreoPaths Load path commands using choreo trajectories
   * @param mirror Should the paths be mirrored
   * @return a command built from the JSON object
   * @throws IOException if attempting to load a path file that does not exist or cannot be read
   * @throws ParseException If attempting to load a path with JSON that cannot be parsed
   */
  public static Command commandFromJson(
      JSONObject commandJson, boolean loadChoreoPaths, boolean mirror)
      throws IOException, ParseException {
    String type = (String) commandJson.get("type");
    JSONObject data = (JSONObject) commandJson.get("data");

    return switch (type) {
      case "wait" -> waitCommandFromData(data);
      case "named" -> namedCommandFromData(data);
      case "path" -> pathCommandFromData(data, loadChoreoPaths, mirror);
      case "sequential" -> sequentialGroupFromData(data, loadChoreoPaths, mirror);
      case "parallel" -> parallelGroupFromData(data, loadChoreoPaths, mirror);
      case "race" -> raceGroupFromData(data, loadChoreoPaths, mirror);
      case "deadline" -> deadlineGroupFromData(data, loadChoreoPaths, mirror);
      default -> Commands.none();
    };
  }

  private static Command waitCommandFromData(JSONObject dataJson) {
    try {
      double waitTime = ((Number) dataJson.get("waitTime")).doubleValue();
      return Commands.waitSeconds(waitTime);
    } catch (Exception ignored) {
      // Failed to load wait time as a number. This is probably a choreo expression
      JSONObject waitTimeJson = (JSONObject) dataJson.get("waitTime");
      double waitTime = ((Number) waitTimeJson.get("val")).doubleValue();
      return Commands.waitSeconds(waitTime);
    }
  }

  private static Command namedCommandFromData(JSONObject dataJson) {
    String name = (String) dataJson.get("name");
    return NamedCommands.getCommand(name);
  }

  private static Command pathCommandFromData(
      JSONObject dataJson, boolean choreoPath, boolean mirror) throws IOException, ParseException {
    String pathName = (String) dataJson.get("pathName");

    PathPlannerPath path =
        choreoPath
            ? PathPlannerPath.fromChoreoTrajectory(pathName)
            : PathPlannerPath.fromPathFile(pathName);
    if (mirror) {
      path = path.mirrorPath();
    }
    return AutoBuilder.followPath(path);
  }

  private static Command sequentialGroupFromData(
      JSONObject dataJson, boolean loadChoreoPaths, boolean mirror)
      throws IOException, ParseException {
    SequentialCommandGroup group = new SequentialCommandGroup();
    for (var cmdJson : (JSONArray) dataJson.get("commands")) {
      group.addCommands(commandFromJson((JSONObject) cmdJson, loadChoreoPaths, mirror));
    }
    return group;
  }

  private static Command parallelGroupFromData(
      JSONObject dataJson, boolean loadChoreoPaths, boolean mirror)
      throws IOException, ParseException {
    ParallelCommandGroup group = new ParallelCommandGroup();
    for (var cmdJson : (JSONArray) dataJson.get("commands")) {
      group.addCommands(commandFromJson((JSONObject) cmdJson, loadChoreoPaths, mirror));
    }
    return group;
  }

  private static Command raceGroupFromData(
      JSONObject dataJson, boolean loadChoreoPaths, boolean mirror)
      throws IOException, ParseException {
    ParallelRaceGroup group = new ParallelRaceGroup();
    for (var cmdJson : (JSONArray) dataJson.get("commands")) {
      group.addCommands(commandFromJson((JSONObject) cmdJson, loadChoreoPaths, mirror));
    }
    return group;
  }

  private static Command deadlineGroupFromData(
      JSONObject dataJson, boolean loadChoreoPaths, boolean mirror)
      throws IOException, ParseException {
    JSONArray cmds = (JSONArray) dataJson.get("commands");

    if (!cmds.isEmpty()) {
      Command deadline = commandFromJson((JSONObject) cmds.get(0), loadChoreoPaths, mirror);
      ParallelDeadlineGroup group = new ParallelDeadlineGroup(deadline);
      for (int i = 1; i < cmds.size(); i++) {
        group.addCommands(commandFromJson((JSONObject) cmds.get(i), loadChoreoPaths, mirror));
      }
      return group;
    } else {
      return Commands.none();
    }
  }
}

File: pathplannerlib/src/main/native/cpp/pathplanner/lib/auto/CommandUtil.cpp
#include "pathplanner/lib/auto/CommandUtil.h"
#include "pathplanner/lib/auto/NamedCommands.h"
#include "pathplanner/lib/auto/AutoBuilder.h"
#include <wpi/commands2/Commands.hpp>
#include <string>
#include <wpi/units/time.hpp>
#include <vector>

using namespace pathplanner;

wpi::cmd::CommandPtr CommandUtil::wrappedEventCommand(
		std::shared_ptr<wpi::cmd::Command> command) {
	wpi::cmd::FunctionalCommand wrapped([command]() {
		command->Initialize();
	},
	[command]() {
		command->Execute();
	},
	[command](bool interrupted) {
		command->End(interrupted);
	},
	[command]() {
		return command->IsFinished();
	}
	);
	wrapped.AddRequirements(command->GetRequirements());

	return std::move(wrapped).ToPtr();
}

wpi::cmd::CommandPtr CommandUtil::commandFromJson(const wpi::util::json &json,
		bool loadChoreoPaths, bool mirror) {
	std::string type = json.at("type").get_string();
	const auto &data = json.at("data");

	if (type == "wait") {
		return CommandUtil::waitCommandFromJson(data);
	} else if (type == "named") {
		return CommandUtil::namedCommandFromJson(data);
	} else if (type == "path") {
		return CommandUtil::pathCommandFromJson(data, loadChoreoPaths, mirror);
	} else if (type == "sequential") {
		return CommandUtil::sequentialGroupFromJson(data, loadChoreoPaths,
				mirror);
	} else if (type == "parallel") {
		return CommandUtil::parallelGroupFromJson(data, loadChoreoPaths, mirror);
	} else if (type == "race") {
		return CommandUtil::raceGroupFromJson(data, loadChoreoPaths, mirror);
	} else if (type == "deadline") {
		return CommandUtil::deadlineGroupFromJson(data, loadChoreoPaths, mirror);
	}

	return wpi::cmd::None();
}

wpi::cmd::CommandPtr CommandUtil::waitCommandFromJson(
		const wpi::util::json &json) {
	auto waitJson = json.at("waitTime");
	if (waitJson.is_number()) {
		return wpi::cmd::Wait(wpi::units::second_t { waitJson.get_number() });
	} else {
		// Field is not a number, probably a choreo expression
		return wpi::cmd::Wait(
				wpi::units::second_t { waitJson.at("val").get_number() });
	}
}

wpi::cmd::CommandPtr CommandUtil::namedCommandFromJson(
		const wpi::util::json &json) {
	std::string name = json.at("name").get_string();
	return NamedCommands::getCommand(name);
}

wpi::cmd::CommandPtr CommandUtil::pathCommandFromJson(
		const wpi::util::json &json, bool loadChoreoPaths, bool mirror) {
	std::string pathName = json.at("pathName").get_string();

	std::shared_ptr < PathPlannerPath > path =
			loadChoreoPaths ?
					PathPlannerPath::fromChoreoTrajectory(pathName) :
					PathPlannerPath::fromPathFile(pathName);

	if (mirror) {
		path = path->mirrorPath();
	}

	return AutoBuilder::followPath(path);
}

wpi::cmd::CommandPtr CommandUtil::sequentialGroupFromJson(
		const wpi::util::json &json, bool loadChoreoPaths, bool mirror) {
	std::vector < wpi::cmd::CommandPtr > commands;

	const auto &commandsJson = json.at("commands").get_array();
	for (size_t i = 0; i < commandsJson.size(); i++) {
		commands.push_back(
				CommandUtil::commandFromJson(commandsJson[i], loadChoreoPaths,
						mirror));
	}

	return wpi::cmd::Sequence(std::move(commands));
}

wpi::cmd::CommandPtr CommandUtil::parallelGroupFromJson(
		const wpi::util::json &json, bool loadChoreoPaths, bool mirror) {
	std::vector < wpi::cmd::CommandPtr > commands;

	const auto &commandsJson = json.at("commands").get_array();
	for (size_t i = 0; i < commandsJson.size(); i++) {
		commands.push_back(
				CommandUtil::commandFromJson(commandsJson[i], loadChoreoPaths,
						mirror));
	}

	return wpi::cmd::Parallel(std::move(commands));
}

wpi::cmd::CommandPtr CommandUtil::raceGroupFromJson(const wpi::util::json &json,
		bool loadChoreoPaths, bool mirror) {
	std::vector < wpi::cmd::CommandPtr > commands;

	const auto &commandsJson = json.at("commands").get_array();
	for (size_t i = 0; i < commandsJson.size(); i++) {
		commands.push_back(
				CommandUtil::commandFromJson(commandsJson[i], loadChoreoPaths,
						mirror));
	}

	return wpi::cmd::Race(std::move(commands));
}

wpi::cmd::CommandPtr CommandUtil::deadlineGroupFromJson(
		const wpi::util::json &json, bool loadChoreoPaths, bool mirror) {
	const auto &commandsJson = json.at("commands").get_array();

	if (commandsJson.size() == 0) {
		return wpi::cmd::None();
	}

	wpi::cmd::CommandPtr deadline = CommandUtil::commandFromJson(
			commandsJson[0], loadChoreoPaths, mirror);
	std::vector < wpi::cmd::CommandPtr > commands;

	for (size_t i = 1; i < commandsJson.size(); i++) {
		commands.push_back(
				CommandUtil::commandFromJson(commandsJson[i], loadChoreoPaths,
						mirror));
	}

	return wpi::cmd::Deadline(std::move(deadline), std::move(commands));
}
